samples/basic-approval-workflow/src/pubnub-client.js

```diff
@@ -0,0 +1,85 @@
+const channelName = (prefix, entrySys) => [
+  prefix,
+  entrySys.space.sys.id,
+  entrySys.environment.sys.id,
+  entrySys.id,
+].join('!');
+
+export default async function createPubNubClient({
+  PubNub,
+  publishKey,
+  subscribeKey,
+  channelPrefix,
+  entrySys,
+  onChange,
+}) {
+  const pubnub = new PubNub({ publishKey, subscribeKey, ssl: true });
+  const channel = channelName(channelPrefix, entrySys);
+  const state = { log: [] };
+
+  function addToLog({ entry, timetoken }) {
+    const validEntry = entry && typeof entry.id === 'string';
+    const validTimetoken = typeof timetoken === 'string';
+    const newEntry = !state.log.find(cur => cur.id === entry.id);
+
+
+    if (validEntry && validTimetoken && newEntry) {
+      // Timetoken is a 17-digit precision unix time (UTC)
+      const t = parseInt(timetoken, 10) / 10000;
+      const item = { ...entry, t };
+      state.log = state.log.concat([item]);
+      state.log.sort((a, b) => b.t - a.t);
+      return true;
+    }
+    return false;
+  }
+
+  function getHistory() {
+    return new Promise((resolve, reject) => {
+      pubnub.history({
+        channel,
+        count: 25,
+        stringifiedTimeToken: true,
+      }, (status, res) => {
+        if (status && status.error) {
+          reject(new Error('Failed to get channel history.'));
+        } else {
+          resolve(res);
+        }
+      });
+    });
+  }
+
+  function publish(message) {
+    return new Promise((resolve, reject) => {
+      pubnub.publish({ channel, message }, (status, res) => {
```

Comment:
Same, here, [there's a Promise version for publish as well](https://www.pubnub.com/docs/web-javascript/api-reference-publish-and-subscribe#publish-basic-usage-using-promises)